ユーザー管理機能

sorairo580 opened the pull request:
# What
〇ユーザー管理機能ブランチを作成
〇deviseの導入
〇ユーザー管理機能の実装
会員登録動画↓↓
https://gyazo.com/daa9bd6bee301963d964ab48a293a5e4
会員登録エラー時の動画↓↓
https://gyazo.com/db1b6ab24e8f54bba0592c4f4b8ba59d
ログインの動画↓↓
https://gyazo.com/7b0da611e25590923958a260738f60d8
ログインエラー時の動画↓↓
https://gyazo.com/f4c7a6c824d05169787da6eff8098e34
https://gyazo.com/6f0eb2386e1ed293f724230869871df9
ログアウト時の動画↓↓
https://gyazo.com/186aea94d24aa998cba381ac7ef4d64c
〇テストコードの記載
テスト時画像↓↓
https://gyazo.com/f619b537c17608b4f4ebfc524c469746

# Why
ユーザー管理機能の実装のため

sorairo580 committed:
ユーザー管理機能１

ブランチを作成、Deviseを導入、ユーザー管理機能を実装のうちバリデーション（新規登録/ユーザー情報、新規登録/本人情報確認）、トップページのヘッダーに「新規登録」「ログイン」ボタン

sorairo580 committed:
データベースの再構築

encrypted_passwordをpasswordへすべて変更
birthdayカラムが追加されていなかったため再度追加
トップページがサインイン画面になってしまったので修正予定

sorairo580 committed:
アクセス時のトップページ修正

before_action :authenticate_user!をコメントアウト

sorairo580 committed:
ユーザーの新規登録

DBeverのカラム名のみencrypted_passwordに最修正
birthdayカラムがはじかれたが、無事保存された（理由不明）

sorairo580 committed:
ログインの有無での表示変更

sorairo580 committed:
ログイン機能及び新規登録ページの実装

ログイン時に正しいアドレスとパスワードを入れても入れない

sorairo580 committed:
ログイン機能の実装

sorairo580 committed:
rspecの導入

sorairo580 committed:
userモデル単体テストコードの実装

不用なコメントアウトの削除
テストコード記述
Rubocopの実行

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+         :recoverable, :rememberable, :validatable
+
+  validates :nickname, presence: true
+  validates :email,    presence: true, uniqueness: { case_sensitive: true },
```
```rb
  validates :email,    presence: true, uniqueness: { case_sensitive: true },
                       format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
```
emailへのバリデーションはdeviseのデフォルトの機能ですでに実装されているので、削除しましょう。

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  validates :email,    presence: true, uniqueness: { case_sensitive: true },
+                       format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
+  validates :password, presence: true, format: { with: /\A(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]+\z/ }, length: { minimum: 6 }
+  validates :password_confirmation, presence: true
```
```rb
validates :password_confirmation, presence: true
```
こちらもdeviseに標準で設定されているものなので、削除しましょう。

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  validates :nickname, presence: true
+  validates :email,    presence: true, uniqueness: { case_sensitive: true },
+                       format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
+  validates :password, presence: true, format: { with: /\A(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]+\z/ }, length: { minimum: 6 }
```
`presence: true,`の部分はデフォルトで設定されているので、削除しましょう。

TakashiKozuma commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  validates :password, presence: true, format: { with: /\A(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]+\z/ }, length: { minimum: 6 }
+  validates :password_confirmation, presence: true
+  validates :last_name, presence: true, format: { with: /\A[\p{Hiragana}\p{Katakana}\p{Han}ー－]+\z/u }
+  validates :first_name, presence: true, format: { with: /\A[\p{Hiragana}\p{Katakana}\p{Han}ー－]+\z/u }
```
```rb
  validates :last_name, presence: true, format: { with: /\A[\p{Hiragana}\p{Katakana}\p{Han}ー－]+\z/u }
  validates :first_name, presence: true, format: { with: /\A[\p{Hiragana}\p{Katakana}\p{Han}ー－]+\z/u }
```
現状の正規表現だと、「市ヶ谷」などの「ヶ」が登録できないため、正規表現を修正しましょう。

TakashiKozuma commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
+  factory :user do
+    nickname              { 'test' }
+    email                 { 'test@example' }
```
```rb
email { 'test@example' }
```
emailは一意制約を持つカラムなので、Fakerを用いてランダムで生成されるように修正しましょう。

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,59 @@
+  before do
+    @user = FactoryBot.build(:user)
+  end
+
```
以下のようにcontextを使用して正常系と異常系とを分けて記述しましょう！
そうすることで、とても可読性の高いコードとなります。

context '新規登録ができる時' do
  it '' do
  end
end
context '新規登録ができない時' do
  it '' do
  end
end

TakashiKozuma commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,59 @@
+      @user.valid?
+      expect(@user.errors.full_messages).to include("Birthday can't be blank")
+    end
+  end
```
下記の条件を確認するためのテストも追加しましょう。
- メールアドレスが一意性であること
- メールアドレスは、@を含む必要があること
- パスワードは、6文字以上での入力が必須であること
- メールアドレスが一意性であること
メールアドレスは、@を含む必要があること
- お名前(全角)は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
- お名前カナ(全角)は、全角（カタカナ）での入力が必須であること

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
コメントを致しましたので、ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

sorairo580 committed:
正規表現の修正

名前に「ヶ」まで含まれるよう修正

sorairo580 committed:
Faker及びFaker-japaneseの導入

sorairo580 committed:
テストコードの修正

メールアドレスが一意性であること
メールアドレスは、@を含む必要があること
パスワードは、6文字以上での入力が必須であること
お名前(全角)は、全角（漢字・ひらがな・カタカナ）での入力が必須であること
お名前カナ(全角)は、全角（カタカナ）での入力が必須であること

sorairo580 commented:
修正しましたので、ご確認をお願いします。

正規表現の修正↓↓
https://gyazo.com/7a97e9a6692bd805b5b4cb9605dca614
Faker及びFaker-japaneseの導入↓↓
https://gyazo.com/45c33674d920254f146cb9708a125d2f
テストコードの修正↓↓
https://gyazo.com/211bc210a7ff4f3b9cc74f31a261d61c
https://gyazo.com/6b5b56d96e30d0c983dd43a47535ac5f
https://gyazo.com/6672362548df2a4b264b99f7099c403a
https://gyazo.com/c0e0e63cc2c5136ab9d2b06c77ad5648
ターミナルの画像↓↓
https://gyazo.com/81391282685e7cae6f7b7d71f81496ba
https://gyazo.com/866444055f2bd81bd85c841dd325013d
https://gyazo.com/aec3ba7354cdbdb3783b77da0aef3381

sorairo580 committed:
rubocopによる修正

minachanchan commented on README.md:
```diff
@@ -5,7 +5,7 @@
 |--------------------|--------|---------------------------|
 | email              | string | null: false, unique: true |
-| encrypted_password | string | null: false               |
+| password           | string | null: false               |
```
```ruby
| password | string | null: false |
```
こちらは最終的にパスワードの値を保存することになるカラム としてencrypted_passwordに戻しましょう。

minachanchan commented on app/models/user.rb:
```diff
@@ -0,0 +1,14 @@
+         :recoverable, :rememberable, :validatable
+
+  validates :nickname, presence: true
+  validates :password, format: { with: /\A(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]+\z/ }, length: { minimum: 6 }
```
```ruby
validates :password, format: { with: /\A(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]+\z/ }, length: { minimum: 6 }
```
文字数についてのバリデーションの指定は不要ですので `length: { minimum: 6 }`を削除しましょう！

minachanchan commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,34 @@
+<%= render "shared/second-header"%>
+
+<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
```erb
<%= form_with model: @user, url: new_user_session_path, class: 'registration-main', local: true do |f| %>
```
こちらで指定しているパスを見直しましょう！
`new_user_session_path`はログインページへアクセスする際のsessions#newへのパスです。
sessions#createへのパスを指定しましょう！

minachanchan commented on db/migrate/20230916055924_add_birthday_to_users.rb:
```diff
@@ -0,0 +1,5 @@
+class AddBirthdayToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :birthday, :date
```
```ruby
add_column :users, :birthday, :date
```
こちらのカラムにもnull: falseを指定しましょう！

minachanchan commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              {Faker::Name.initials(number: 6)}
+    email                 {Faker::Internet.free_email}
+    password              {Faker::Internet.password(min_length: 6)}
```
```ruby
password {Faker::Internet.password(min_length: 6)}
```

Fakerでのpasswordの生成について記述できていて素晴らしいです！
こちらの記述ですと、稀に英字のみ、もしくは数値のみの値が生成されることがございます。
以下のように、文字列として英数字を予め含めておきましょう！

```ruby
password              {'1a' + Faker::Internet.password(min_length: 6)}
```

minachanchan commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,125 @@
+        @user.password_confirmation = '123za'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password is too short (minimum is 6 characters)")
+      end
```
passwordに関しては以下の項目について確認するテストコードを記述しましょう！

```
・英字のみでは登録できないこと
・数字のみでは登録できないこと
・全角だと登録できないこと
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
着実に完成に近づいておりますね！

コメントいたしましたので、ご確認頂けますと幸いでございます。


実装完了まで引き続き、一緒に頑張りましょう！
修正の後、再度レビューのご依頼頂けますようお願い申し上げます。

sorairo580 committed:
ユーザー管理機能修正２

READMEの修正
文字数のバリデーションの削除
sessions#createへのパスの変更
birthdayカラムのバリデーション追加
Fakerにおけるpassword作成時の対策
passwordのテスト項目追加

sorairo580 commented:
ご確認ありがとうございます。
修正しましたので、ご確認をお願いします。

・READMEの修正
・文字数のバリデーションの削除
・sessions#createへのパスの変更
・birthdayカラムのバリデーション追加
・Fakerにおけるpassword作成時の対策
・passwordのテスト項目追加

READMEの修正↓↓
https://gyazo.com/1c787347118668dce34c8082b6409dcf

AsukaToyoshima commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,106 @@
+    </h1>
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
こちらのコメントアウトは削除しておきましょう。
実装を指示するコメントが残っている状態は望ましくありません。
他にも同様の箇所がありますので、併せて削除ください。

AsukaToyoshima commented on app/views/shared/_header.html.erb:
```diff
@@ -16,10 +16,13 @@
     </ul>
     <ul class='lists-right'>
       <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
こちらのコメントアウトは削除しておきましょう。
実装を指示するコメントが残っている状態は望ましくありません。
他にも同様の箇所がありますので、併せて削除ください。

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
修正点につきまして、確認させていただきました。
大変恐縮ではございますが、追加でコメントを付けさせていただきました。ご確認の上、修正をお願いいたします。
疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

sorairo580 committed:
不要なコメントアウトの削除

sorairo580 commented:
ご確認ありがとうございます。

不要なコメントアウト部分は削除しましたので、再度ご確認をお願いします。

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
こちらLGTMとさせていただきます。お疲れ様でした！
引き続き、実装頑張っていきましょう！　応援しております！

sorairo580 commented:
ありがとうございます！
がんばります！
